src/cryptonote_core/blockchain.cpp

```diff
@@ -3611,7 +3577,8 @@ bool Blockchain::cleanup_handle_incoming_blocks(bool force_sync)
       if(m_db_sync_mode == db_async)
       {
         m_sync_counter = 0;
-        m_async_service.dispatch(boost::bind(&Blockchain::store_blockchain, this));
+        tools::threadpool& tpool = tools::threadpool::getInstance();
+        tpool.submit(NULL, boost::bind(&Blockchain::store_blockchain, this));
```

Comment:
The semantics change now. Previously `m_async_service` was always destructed before `this`, and now `this` can be destructed after `tpool`. The obvious "fix" would be a waiter in place of `m_async_service`.

Reply:
I didn't state this exactly correct. The problem is that `this` can be destructed before the submitted tasks is complete without the waiter. That was the academic rationale for the `task_region` in the C++ literature I saw, it made it more difficult to leaving dangling references.

Reply:
Yeah, I see.

Reply:
And it just occurred to me that this could be one of those interesting "blocking" places. This call does I/O to some user specified filesystem. Since there is no "working-stealing" in this design (possibly a good thing), the wait calls in the range signature verification will have to wait for I/O depending on various factors (number of threads in the pool, and number of times stored is called). I recall this being why I went with the simple create-on-use upgrade first. Perhaps instead the I/O calls should just continue to use their own thread ?

Reply:
Good point. Yeah, perhaps this particular chunk should be left as it was originally.